api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -16,6 +19,80 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: Unified Matcher API]
 
+message MatchTree {
+  message MatchAction {
+    oneof action {
+      bool skip = 1;
+
+      string callback = 2;
+    }
+  }
+
+  message MatchLeaf {
+    message LeafMatcher {
+      // The predicate to evaluate.
+      oneof matcher {
+        MatchPredicate predicate = 1;
```

Comment:
Where is this defined?

Reply:
https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/common/matcher/v3/matcher.proto

Reply:
Can we better document which predicates are supported in which situations? This is where I'm still confused as to how this is going to play out. Do we need to split that message up into its component types to make more sense?

Reply:
Formalizing what I said about this earlier a bit more I think it would be better to think about restricting input data for a match rather than predicates: for example, instead of saying you can only use RequestHeaderMatcher we say that you can only use matchers that require request headers. This would open up for custom matchers to be added to MatchPredicate that consume request headers without having to update filters to explicitly list all the supported matchers. 

Validation would then work by traversing the match tree and comparing the requirements for each matcher with the allowed requirements for the filter/match context and reject the config if there is a mismatch. 

While I haven't proved out that code I believe we can proceed with this API and add in this validation on top of it. LMK if you think otherwise and I will try to prove it out before starting to break this PR up.

Reply:
I think this sounds good, but I'm not sure what the win is from the perspective of filters. In both cases (match criteria and predicate `oneof`), there would be some library that takes the matcher proto and hides the implementation busyness from the filter.

Also, do we really want matchers that specialize on request headers? I think we were in general agreement on making this key/value bag matching, which would work for headers or anything else.

Reply:
The idea would be to let a filter influence the maximum amount of required input for a match in order to provide guarantees around when the match result will be ready: if a filter needs to know the match result before decoding headers, it could specify that the match tree needs to only contain matchers that consume request headers. 

For specialized matchers I think I was mostly thinking about extensions, e.g. if someone wants to add their own matcher that consumes request and response headers, being able to advertise that it does so would prevent users from using it with filters that require a matching result early on.

One thing to note here is that we're still using the MatchPredicate which uses HeaderMatcher - maybe we should be moving that out in favor of a more generic matcher here as well?

Reply:
Probably. Will continue discussion in your dedicate PR.